_drafts/fs2.md

```diff
@@ -0,0 +1,722 @@
+```
+
+In the above example, we react to an error by returning a stream that prints the error to the console. As we can notice, the elements contained in the stream are `AnyVal` and not `Unit` because of the definition of the `handleErrorWith`:
+
+```scala
+// fs2 library code
+def handleErrorWith[F2[x] >: F[x], O2 >: O](h: Throwable => Stream[F2, O2]): Stream[F2, O2] = ???
+```
+
+The `O2` type must be a supertype of `O`'s original type. Since both `Int` and `Unit` are subtypes of `AnyVal`, we can use the `AnyVal` type (the least common supertype) to represent the resulting stream.
+
+Another available method to handle errors in streams is `attempt`. The method works using the `scala.Either` type, which returns a stream of `Either` elements. The resulting stream pulls elements wrapped in a `Right` instance until the first error occurs, which is nothing more than an instance of a `Throwable` wrapped in a `Left`:
+
+```scala
+val attemptedSavedJlActors: Stream[IO, Either[Throwable, Int]] = savedJlActors.attempt
+attemptedSavedJlActors.evalMap {
+  case Left(error) => IO.println(s"Error: $error")
+  case Right(id) => IO.println(s"Saved actor with id: $id")
+}
+```
+
+For the sake of completeness, let's say that the `attempt` method is implemented internally using the `handleErrorWith` most straightforwardly:
+
+```scala
+// fs2 library code
+def attempt: Stream[F, Either[Throwable, O]] =
+  map(Right(_): Either[Throwable, O]).handleErrorWith(e => Stream.emit(Left(e)))
+```
+
+## 5. Resource Management
+
+As the official documentation says, we don't have to use the `handleErrorWith` method to manage the release of resources used by the stream eventually. Instead, **the fs2 library implements the _bracket_ pattern to manage resources**.
+
+The bracket pattern is a resource management pattern developed in the functional programming domain many years ago. The pattern defines two functions: The first is used to acquire a resource; The second is guaranteed to be called when the resource is no longer needed.
+
+The fs2 library implements the bracket pattern through the `bracket` method:
+
+```scala
+// fs2 library code
+def bracket[F[x] >: Pure[x], R](acquire: F[R])(release: R => F[Unit]): Stream[F, R] = ???
+```
+
+The function `acquire` is used to acquire a resource, and the function `release` is used to release the resource. The resulting stream has elements of type `R`. Moreover, **the resource is released at the end, both in case of success and in case of error**. Notice that both the `acquire` and `release` functions returns an effect since they can throw exceptions during the acquisition or release of resources.
+
+Let's make an example. We want to use a resource during the persistence of the stream containing the actors of the JLA. First, we define a value class representing a connection to a database:
+
+```scala
+case class DatabaseConnection(connection: String) extends AnyVal
```

Comment:
why `extends AnyVal`?

Reply:
Because I want to create a value class, however, it's not the article's primary focus. So, I can remove it.